Add support for Battlefield: Bad Company 2

cetteup opened the pull request:
Added support for reading player name and prefix for Battlefield: Bad Company 2.

## Changes
- Add reader class and interface for Battlefield: Bad Company 2
- Update class/namespace structure
- Use common code to parse line based config files based on regular expressions

cetteup committed:
feat: Add support for reading Battlefield: Bad Company 2 player name and prefix

cetteup committed:
refactor: Simplify Refractor config file by extending LineBasedConfigFile

cetteup committed:
chore: Remove now obsolete RefractorCommon parsing

cetteup committed:
refactor: Use GetValue instead of ParseValue since not all settings need parsing

cetteup committed:
refactor: Move all common classes to Common namespace

cetteup committed:
refactor: Introduce constants for common regex group names

cetteup committed:
chore: Extend offered service registrations for DI

TwitchPlaysBF2 commented on src/GameDataReader/BattlefieldBadCompany2/Parsing/BfBc2GameSettingsBinSettingResolver.cs:
```diff
@@ -0,0 +1,46 @@
+         * a pattern of [key]...stuff...[value] emerges. We need to ignore some key for which there no (readable) values,
+         * such as FlashValues, UIMenuTrackerPage_Store, UIMenuTrackerPage_MenuUnlocks etc.
+         */
+        var re = new Regex($@"(?<{Constants.RegexKeyGroupName}>[a-zA-Z][\w]+)(?<!UIMenu\w+|FlashValues)[\x00-\x1F\x7f-\x9f\u2122\ufffd\s]+(?<{Constants.RegexValueGroupName}>[\x20-\x7e]+)");
```
Never use abbreviations as variable names please - better: "regex" or something alike (agreed it's pretty dull in this case)

Reason might be bit personal but it's been an unwritten rule violation in all my previous dotnet work environments. Mainly to enforce easier readability I guess. There are a few exceptions:
- i and j as iterators in a for loop
- e or ex for `catch(Exception ex)`
- e for `EventArgs e`
- Lamba expression parameters i.e. `Repository.GetUsers().Where(u => u.IsActive);`

TwitchPlaysBF2 committed:
Refactor abstract classes into interfaces:

	- Abstract classes without behavior should always be interfaces
	- Implementing public within an internal class is ok since the class isn't exposed anyway

TwitchPlaysBF2 committed:
Minor renaming

TwitchPlaysBF2 committed:
Rename classes (shorten names to improve overview)

TwitchPlaysBF2 committed:
Cleanup comment

TwitchPlaysBF2 committed:
Update README.md

TwitchPlaysBF2 commented on src/GameDataReader/Common/Parsing/Setting.cs:
```diff
@@ -0,0 +1,6 @@
+﻿namespace GameDataReader.Common.Parsing;
+
+internal abstract class Setting
```
Abstract classes that don't have behavior should be interfaces (there are probably exceptions if you wanna pipe some abstraction layers but this is the last layer so we want it here)

(I pushed the changes for that already - also for all the other similar classes)

TwitchPlaysBF2 commented on src/GameDataReader/GameDataReaders.cs:
```diff
@@ -11,9 +12,11 @@ public static class GameDataReaders
     public static IBf1942DataReader Bf1942 => Bf1942Singleton.Value;
     public static IBf2DataReader Bf2 => Bf2Singleton.Value;
     public static IBf2142DataReader Bf2142 => Bf2142Singleton.Value;
+    public static IBfBc2DataReader BfBc2 => BfBc2Singleton.Value;
```
Readme was missing (I quickly extended it)

TwitchPlaysBF2 reviewed (approved):
Really solid extension!

I'm overall pleased with how well you structured the bin/ini stuff. But my initial structure kind of got into the way of structuring this nicer. Especially when it comes to testing which is really needed, because now I'm too afraid of refactoring anything. 

I liked how you extracted things to the Common namespace.

👍
